src/jupyterlabcontroller/services/fileserver.py

```diff
@@ -0,0 +1,422 @@
+from ..models.v1.lab import UserInfo
+from ..storage.k8s import K8sStorageClient
+
+
+class FileserverReconciler:
+    def __init__(
+        self,
+        *,
+        config: Config,
+        user_map: FileserverUserMap,
+        logger: BoundLogger,
+        k8s_client: K8sStorageClient,
+    ) -> None:
+        self.fs_namespace = config.fileserver.namespace
+        self.user_map = user_map
+        self.k8s_client = k8s_client
+        self._logger = logger
+        self._tasks: Set[asyncio.Task] = set()
+        self._started = False
+
+    async def start(self) -> None:
+        if not await self.k8s_client.check_namespace(self.fs_namespace):
+            self._logger.warning(
+                "No namespace '{self.fs_namespace}'; cannot start"
+            )
+            return
+        if self._started:
+            msg = "Fileserver reconciliation already running; cannot start"
+            self._logger.warning(msg)
+            return
+        self._started = True
+        self._logger.info("Starting fileserver reconciliation")
+        reconciliation_task = asyncio.create_task(self._reconciliation_loop())
+        self._tasks.add(reconciliation_task)
+        reconciliation_task.add_done_callback(self._tasks.discard)
+
+    async def _reconciliation_loop(self) -> None:
+        while self._started:
+            await self.reconcile_user_map()
+            await asyncio.sleep(
+                FILESERVER_RECONCILIATION_INTERVAL.total_seconds()
+            )
+
+    async def stop(self) -> None:
+        if not self._started:
+            msg = "Fileserver reconciliation already stopped"
+            self._logger.warning(msg)
+            return
+        self._started = False
+        for tsk in self._tasks:
+            tsk.cancel()
+
+    async def reconcile_user_map(self) -> None:
+        self._logger.debug("Reconciling fileserver user map")
+        namespace = self.fs_namespace
+        if not await self.k8s_client.check_namespace(namespace):
+            self._logger.warning(
+                f"No fileserver namespace '{namespace}'; no fileserver users"
+            )
+            return
+        mapped_users = set(self.user_map.list_users())
+        observed_map = await self.k8s_client.get_observed_fileserver_state(
+            namespace
+        )
+        # Tidy up any no-longer-running users.  They aren't running, but they
+        # might have some objects remaining.
+        observed_users = set(observed_map.keys())
+        missing_users = mapped_users - observed_users
+        if missing_users:
+            self._logger.info(
+                f"Users {missing_users} have broken fileservers; removing."
+            )
+        for user in missing_users:
+            rmuser_task = asyncio.create_task(
+                self.k8s_client.remove_fileserver(user, namespace)
+            )
+            self._tasks.add(rmuser_task)
+            rmuser_task.add_done_callback(self._tasks.discard)
```

Comment:
Foreground removals it is.

Will think through the concurrency later.

Reply:
I've added locking.  Not sure it's in quite the right place, but basically there's a shared lock that protects access to the user map, and is asserted around fileserver creation and deletion.  I don't expect this to be massively parallel, because this is just the interface a user uses to start a fileserver, and then they ignore the endpoint until they let the server time out and want to do WebDAVy things again...at which point they will notice their client is broken, and go to /files again.  I doubt if we ever even get to more-than-one-a-second in actual use.  (Sure, we'll have lots of people going through created ingresses to their fileservers, but that's the reason Worblehat should be very lightweight.)